Chore/deployment

manvinderjit opened the pull request:
## Summary

Create a GitHub Actions pipeline to automatically deploy the project to dev and production environments.

## Tests

Tested manually in a private fork of the repository

## Additional information

Need to do the following for this to work:

1. Update API tokens and GitHub secrets
2. Create a Cloudflare API user with appropriate permissions.

manvinderjit committed:
chore (deployment): add github workflow for deployment

add github actions workflow for deployment to dev and prod

manvinderjit committed:
chore (deployment): update wrangler toml to use envs

update wrangler.toml to use envs injected from github secrets.

manvinderjit committed:
Merge remote-tracking branch 'origin/dev' into chore/deployment

manvinderjit committed:
Revert "chore(migrations): rename files to cloudflare naming convention"

This reverts commit ca4b143e02f222c4f5ec7e03b3608f299acdda9c.
Restore migration file names to original format.

manvinderjit committed:
chore (seed): update seed data for pbkdf2 100,000

update seed data to change passwords for 100,000 iteration pbkdf2.

madcampos commented on .github/workflows/cd-deployments.yml:
**question**: I think right now the deployment is done "in reverse", from cloudflare observing the changes on github.

This would change it to go from github to cloudflare right?
Do we need to disable the connection on cloudflare's dashboard as well?

manvinderjit replied:
Yes. 

I have disabled the deployments on CloudFlare for the following:

1. Volunteer (Pages-based deployment, doesn't use workers)
2. VMS-Dev (Old branch, before frontend and backend consolidation)
3. VMS (Old branch, before frontend and backend consolidation)

<img width="1007" height="756" alt="image" src="https://github.com/user-attachments/assets/b49d2293-7d31-4e5f-ba3f-fbc21bbff002" />

@madcampos @kbventures Do we want to delete the older workers and pages or keep them for reference?

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+      deployments: write
+
+    env:
+      NODE_VERSION: 22
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Can we set this to a moving target like "lts"?

Or rather, skip hardcoding the version here entirely and relay on the `.nvmrc` file?

manvinderjit replied:
The workflow now users the ```.nvmrc``` file!

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+      PROD_COMM_HUB_SESSION_TOKENS_ID: ${{ secrets.PROD_COMM_HUB_SESSION_TOKENS_ID }}
+      PROD_COMM_HUB_ACTIVATION_TOKENS_ID: ${{ secrets.PROD_COMM_HUB_ACTIVATION_TOKENS_ID }}
+      PROD_COMM_HUB_PASSWORD_RESET_TOKENS_ID: ${{ secrets.PROD_COMM_HUB_PASSWORD_RESET_TOKENS_ID }}
+      PROD_COMM_HUB_SENDER_EMAIL: ${{ secrets.PROD_COMM_HUB_SENDER_EMAIL }}
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Would it be possible to use the variables set on cloudflare's dashboard instead or do we absolutely need to set them here?

That way they are not present on `wrangler.toml` and so are not visible to everyone and also not on github secrets which have known issues with leaking secrets

manvinderjit replied:
These variables are for substitution into the wrangler toml file for bindings. Cloudflare recommends having them in the wrangler file instead of using keep_vars=true and using variables from both places.

madcampos replied:
I don't particularly like string interpolation like this to be honest... It sounds hacky... :/

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+      - name: Set up Node
+        uses: actions/setup-node@v4
+        with:
+          node-version: ${{ env.NODE_VERSION }}
```
Same question here: can we infer it from the nvmrc file?

manvinderjit replied:
Sure.

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+        run: |
+          npm ci
+          sudo apt-get update
+          # Install envsubst and jq to substitute secrets in wrangler.toml
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** wrnagler also supports a `wrangler.json` file as an alternative to the TOML syntax. Maybe using that instead will make things easier, no?

manvinderjit replied:
I already have created an issue around that but the goal is to get it working first and once we have it figured out, we can migrate to jsonc file. 
The jsonc file gives additional configuration options but we still can't use GitHub secrets in it.

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+        run: |
+          npx wrangler d1 create ${{ env.DEV_COMM_HUB_D1_DB_NAME }} --env dev > /dev/null
+          DB_ID=$(npx wrangler d1 info ${{ env.DEV_COMM_HUB_D1_DB_NAME }} --json | jq -r '.uuid')
+          echo "DEV_COMM_HUB_D1_DB_ID=$DB_ID" >> $GITHUB_ENV
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Why delete and recreate the dev db every time?

Would it make sense instead to invoke an SQL script that drops the tables on the database, or run a rollback on all migrations?

manvinderjit replied:
@kbventures wanted to delete and recreate the database each time. 

Also, deleting and recreating the database is faster, compared to using a SQL script that drops the tables on the database or rolling back migrations. It will reduce the reads and writes on the database so that we don't go above the free tier.

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+
+          # Use envsubst to replace DEV placeholders in wrangler.toml, overwriting the file
+          envsubst "$ENV_VARS_TO_SUB" < wrangler.toml > wrangler.toml.tmp
+          mv wrangler.toml.tmp wrangler.toml
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Two things:
1. If the variables are okay to be shared, like a random meaningless id for something, I think it is fine leaving it in the wrangler config.
2. As this is dev, why not add a `.dev.vars` file with all of the values instead of fiddling with the wrangler config?

manvinderjit replied:
1. While it is generally considered acceptable to have the wrangler file public, it may be a better idea to obscure the values if possible.
2. All the ids and configuration values must be present in the wrangler.toml or wrangler.jsonc file. The thing is, there is no way to reference dev.vars or GitHub secrets in the file. So I am editing the file on the fly to substitute the values. 

The other option is to have the wrangler file committed to the repo, making it public. I would prefer to not do that just in case.

madcampos replied:
So, instead of editing the `wrangler.toml` file, why not output the variables from github secrets to a `.dev.vars` file? That way there is no interpolation on the wrangler config, and the secrets are still segregated to a single place.

My concerns are basically with string interpolation and how that looks hacky, and the separation of regular variables from secrets.

madcampos replied:
We should use secrets instead of variables: https://developers.cloudflare.com/workers/configuration/secrets/#compare-secrets-and-environment-variables

Furthermore:
> To set different secrets for each Cloudflare environment, create files named `.dev.vars.<environment-name>` or `.env.<environment-name>`.

[Reference](https://developers.cloudflare.com/workers/configuration/secrets/#:~:text=To%20set%20different%20secrets%20for%20each%20Cloudflare%20environment%2C%20create%20files%20named%20.dev.vars.%3Cenvironment%2Dname%3E%20or%20.env.%3Cenvironment%2Dname%3E.)

madcampos replied:
I've closed the other comments related to this to keep everything in one thread.

manvinderjit replied:
Reverted ```wrangler.toml``` to previous version wherein it has all the variables. Makes the process simpler and is a Cloudflare recommended standard approach.

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+          envsubst "$ENV_VARS_TO_SUB" < wrangler.toml > wrangler.toml.tmp
+          mv wrangler.toml.tmp wrangler.toml
+
+      - name: Apply D1 Migrations to DEV D1 Database (Dev Only)
```
**thought:** Migrations should be applied in all environments. They also need to be thoroughly tested to be compatible with the previous data and not break anything.

Like, if a new version needs a new format for the database, **not** applying the migration will break the application. So it makes sense to apply the migrations regardless of environment.

manvinderjit replied:
Migrations are being applied to both environments. I kept the steps separate to minimize the chances of dev migration accidentally being deployed to prod accidentally.

madcampos replied:
If you are configuring everything based on the environment, I don't see a problem on keeping the same code and using a variable to set the environment.

In fact it is better and less error prone than duplication.

manvinderjit replied:
Migrations are now applied based on the environment using the same block of code.

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+
+          # Use envsubst to replace PROD placeholders in wrangler.toml, overwriting the file
+          envsubst "$ENV_VARS_TO_SUB" < wrangler.toml > wrangler.toml.tmp
+          mv wrangler.toml.tmp wrangler.toml
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** These variables should be kept on cloudflare's dashboard and read from there instead of messing around with the config file

manvinderjit replied:
These variables can NOT be kept on cloudflare's dashboard because they are used for the wrangler file. The file requires the bindings to be present. While defining bindings in the dashboard is possible, but because we are also using CI/CD, the recommendations is to use the wrangler file as source of truth

<img width="760" height="582" alt="image" src="https://github.com/user-attachments/assets/34c80e6a-ba88-4bbb-a0d9-1808354c3830" />

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+        env:
+          PROD_COMM_HUB_D1_DB_ID: ${{ env.PROD_COMM_HUB_D1_DB_ID }}
+        if: ${{ env.TARGET_ENV == 'production' }}
+        run: npx wrangler d1 migrations apply Database --env production --remote
```
This is the same command as the dev one, expect for the environment. use the `TARGET_ENV` variable instead and combine everything into a single action.

manvinderjit replied:
Migrations are being applied to both environments. I kept the steps separate to minimize the chances of dev migration accidentally being deployed to prod accidentally.

madcampos commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,129 @@
+          secrets: |
+            RESEND_API_KEY
+        env:
+          RESEND_API_KEY: ${{ env.TARGET_ENV == 'production' && secrets.PROD_COMM_HUB_RESEND_API_KEY || secrets.DEV_COMM_HUB_RESEND_API_KEY}}
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Please read their docs and check if the secret can be kept in cloudflare's dashboard instead of being passed arround like this.

manvinderjit replied:
This command deploys the key in Cloudflare's Dashboard itself and not in wrangler file. If you want, we can remove it from the pipeline and put it directly in cloudflare's dashboard.

manvinderjit replied:
Resend API key is now in Cloudflare's Dashboard. We will have to manually update it though for future changes.

madcampos commented on db/seeds/seed-data.sql:
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise. -->
**praise:** 10/10 love that the passowrds were kept in theme haha

[Brian](https://www.youtube.com/watch?v=jHPOzQzk9Qo&pp=ygU0bGlmZSBvZiBicmlhbiBhbHdheXMgbG9vayBvbiB0aGUgYnJpZ2h0IHNpZGUgb2YgbGlmZQ%3D%3D) would be pleased.

madcampos commented on wrangler.toml:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Why not keep the ids here?

If that is a security concern, I think the best option is to provide them in two places instead:
1. `.dev.vars` for local development
2. Cloudflare's dashboard for dev and prod environments

We should also set the `keep_vars` option to `true` so they are not deleted when wrangler run and the values will be merged with what is on the dashboard.

manvinderjit replied:
1. Refer https://github.com/torontojs/communityhub/pull/302#discussion_r2550440619

2. While we can merge the values, cloudflare recommends against it. https://github.com/torontojs/communityhub/pull/302#discussion_r2550457397

manvinderjit replied:
ID's  are now in wrangler.toml

manvinderjit committed:
fix(CICD): remove env substitution

remove env variable substitution from wrangler.toml.
revert to hardcoded values in wrangler.toml.

manvinderjit committed:
fix(CI/CD): consolidate migrations to one step

consolidate migrations to a single step for dev and prod.

manvinderjit committed:
Merge remote-tracking branch 'origin/dev' into chore/deployment

manvinderjit committed:
fix(tests): update seed data in tests

update seed data in data.test.ts.
update seed data in index.test.ts.

manvinderjit committed:
fix(auth): change password hashing iterations count

manvinderjit committed:
fix(tests): update hurl api tests vars for new seed data

manvinderjit committed:
fix: enable deployment for chore/deployment

manvinderjit committed:
fix(deploy): delete tables instead of database for dev

manvinderjit committed:
fix(deploy): selection filter for dev table deletions

manvinderjit committed:
fix(deploy): add drop dev tables command

manvinderjit committed:
fix(deploy): force drop dev tables to ignore key constraints

manvinderjit committed:
fix(deploy): reverse tables list to ignore key constraints

manvinderjit committed:
fix(deploy): remove begin/commit for query to ignore key constraints

manvinderjit committed:
fix(deploy): list tables for dev D1 deletion

manvinderjit committed:
fix(deploy): drop dev tables if they exist

manvinderjit committed:
fix(CI/CD): remove resend api key from pipeline

manvinderjit committed:
fix(CI/CD): remove R2 bucket

manvinderjit committed:
fix(CI/CD): clean workflow steps

bcha92 reviewed (dismissed):
Looks good. Approving this. But since I can't test this without a cloudflare account, have you addressed Marco's comments?

Shajansheriff commented on .github/workflows/cd-deployments.yml:
```diff
@@ -0,0 +1,84 @@
+          for table in "${TABLES[@]}"; do
+            echo "Dropping table: $table"
+
+            npx wrangler d1 execute Database --env dev --remote --command "DROP TABLE IF EXISTS \"$table\";"
```
We're executing wrangler d1 execute once per table in a loop. Since each call is a remote request, this adds unnecessary latency to the CI pipeline.

Suggestion: move the DROP statements into a single SQL file and execute them in one wrangler d1 execute --file call. This will reduce network overhead and speed up the workflow.

manvinderjit replied:
@Shajansheriff Thank you for the input. 

The pipeline now uses an SQL file to DROP tables.

manvinderjit committed:
fix: use nvmrc for node version in deploy workflow

manvinderjit committed:
chore: disable uploadFile, exportData for R2 removal

disable uploadFile and exportData for scheduler api due to R2 removal.

manvinderjit committed:
fix(deploy): move DROP statements to SQL file

move DROP statements into a single SQL file to execute them in one file call.

manvinderjit committed:
chore: add separate url for dev pages

add separate url for dev pages at /pages/dev-links/index.html.

manvinderjit committed:
chore: set sign in page as entry point

manvinderjit committed:
Merge remote-tracking branch 'origin/dev' into chore/deployment

manvinderjit committed:
cleanup(deploy): remove trigger for deploy on chore/deploy branch
